Querying a single column with a predicate on the same column will not return any results

The following pseudo code would fail:
```python
reader = Reader([SomeSchema.a], predicate=in_lambda([SomeSchema.a], lambda a: True),...)
reader.fetch()
```
Will never find results.

I think the reason is due to this piece of code in the Reader.py:
```python
# Read remaining columns
other_rows = piece.read(
    open_file_func=lambda filepath: file,
    columns=other_column_names,
    partitions=self._dataset.partitions).to_pandas().to_dict('records')

# Remove rows that were filtered out by the predicate
filtered_other_rows = [row for i, row in enumerate(other_rows) if match_predicate_mask[i]]

# Decode remaining columns
decoded_other_rows = [_decode_row(row, self._schema) for row in filtered_other_rows]

# Merge predicate needed columns with the remaining
all_cols = [_merge_two_dicts(a, b) for a, b in zip(decoded_other_rows, filtered_decoded_predicate_rows)]
```
`other_rows` will be `[]` since no columns were loaded. This will result in `zip(decoded_other_rows, filtered_decoded_predicate_rows)` returning an empty list `[]` even `filtered_decoded_predicate_rows` could be an array of `[True, True,....]`

@selitvin While attempting to reproduce the error with a test case in #129, when I used `in_lambda([SomeSchema.a], lambda a: True),`, I got the error below.  But, I don't suppose that's the same error you had encountered originally?

```
        if invalid_column_names:
            raise ValueError('At least some column names requested by the predicate ({}) '
>                            'are not valid schema names: ({})'.format(', '.join(invalid_column_names),
                                                                       ', '.join(all_schema_names)))
E           TypeError: sequence item 0: expected string, UnischemaField found
```

--- a/petastorm/tests/test_predicates.py
+++ b/petastorm/tests/test_predicates.py
@@ -12,102 +12,117 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import unittest
+import pytest
 
 from petastorm.predicates import in_set, in_intersection, \
     in_negate, in_reduce, in_pseudorandom_split, in_lambda
+from petastorm.reader import Reader
+from petastorm.tests.test_common import TestSchema
+from petastorm.workers_pool.dummy_pool import DummyPool
 
 
-class PredicatesTest(unittest.TestCase):
+@pytest.fixture(scope="session")
+def all_values(request, tmpdir_factory):
+    all_values = set()
+    for i in range(10000):
+        all_values.add('guid_' + str(i))
+    return all_values
 
-    @classmethod
-    def setUpClass(cls):
-        cls.all_values = set()
-        for i in range(10000):
-            cls.all_values.add('guid_' + str(i))
 
-    @classmethod
-    def tearDownClass(cls):
-        pass
+def test_inclusion(all_values):
+    for values in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_XXX'}, {'guid_2'}]:
+        test_predicate = in_set(values, 'volume_guid')
+        included_values = set()
+        for val in all_values:
+            if test_predicate.do_include({'volume_guid': val}):
+                included_values.add(val)
+        assert included_values == all_values.intersection(values)
 
-    def test_inclusion(self):
-        for values in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_XXX'}, {'guid_2'}]:
-            test_predicate = in_set(values, 'volume_guid')
+
+def test_list_inclusion(all_values):
+    for values in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_XXX'}, {'guid_XX'}]:
+        test_predicate = in_intersection(values, 'volume_guid')
+        included = test_predicate.do_include({'volume_guid': list(all_values)})
+        assert included != all_values.intersection(values)
+
+
+def test_custom_function(all_values):
+    for value in ['guid_2', 'guid_1', 'guid_5', 'guid_XXX', 'guid_XX']:
+        test_predicate = in_lambda(['volume_guids'], lambda volume_guids, val=value: val in volume_guids)
+        included = test_predicate.do_include({'volume_guids': all_values})
+        assert included == (value in all_values)
+
+
+def test_custom_function_with_state(all_values):
+    counter = [0]
+
+    def pred_func(volume_guids, cntr):
+        cntr[0] += 1
+        return volume_guids in all_values
+
+    test_predicate = in_lambda(['volume_guids'], pred_func, counter)
+    for value in ['guid_2', 'guid_1', 'guid_5', 'guid_XXX', 'guid_XX']:
+        included = test_predicate.do_include({'volume_guids': value})
+        assert included == (value in all_values)
+    assert counter[0] == 5
+
+
+def test_negation(all_values):
+    for values in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_9'}, {'guid_2'}]:
+        test_predicate = in_negate(in_set(values, 'volume_guid'))
+        included_values = set()
+        for val in all_values:
+            if test_predicate.do_include({'volume_guid': val}):
+                included_values.add(val)
+        assert included_values == all_values.difference(values)
+
+
+def test_and_argegarion(all_values):
+    for values1 in [{'guid_0', 'guid_1'}, {'guid_3', 'guid_6', 'guid_20'}, {'guid_2'}]:
+        for values2 in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_9'}, {'guid_2'}]:
+            test_predicate = in_reduce(
+                [in_set(values1, 'volume_guid'), in_set(values2, 'volume_guid')], all)
             included_values = set()
-            for val in PredicatesTest.all_values:
+            for val in all_values:
                 if test_predicate.do_include({'volume_guid': val}):
                     included_values.add(val)
-            self.assertEqual(included_values, PredicatesTest.all_values.intersection(values))
+            assert included_values == values1.intersection(values2)
 
-    def test_list_inclusion(self):
-        for values in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_XXX'}, {'guid_XX'}]:
-            test_predicate = in_intersection(values, 'volume_guid')
-            included = test_predicate.do_include({'volume_guid': list(PredicatesTest.all_values)})
-            self.assertEqual(included, len(PredicatesTest.all_values.intersection(values)) > 0)
 
-    def test_custom_function(self):
-        for value in ['guid_2', 'guid_1', 'guid_5', 'guid_XXX', 'guid_XX']:
-            test_predicate = in_lambda(['volume_guids'], lambda volume_guids, val=value: val in volume_guids)
-            included = test_predicate.do_include({'volume_guids': PredicatesTest.all_values})
-            self.assertEqual(included, value in PredicatesTest.all_values)
-
-    def test_custom_function_with_state(self):
-        counter = [0]
-
-        def pred_func(volume_guids, cntr):
-            cntr[0] += 1
-            return volume_guids in PredicatesTest.all_values
-
-        test_predicate = in_lambda(['volume_guids'], pred_func, counter)
-        for value in ['guid_2', 'guid_1', 'guid_5', 'guid_XXX', 'guid_XX']:
-            included = test_predicate.do_include({'volume_guids': value})
-            self.assertEqual(included, value in PredicatesTest.all_values)
-        self.assertEqual(counter[0], 5)
-
-    def test_negation(self):
-        for values in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_9'}, {'guid_2'}]:
-            test_predicate = in_negate(in_set(values, 'volume_guid'))
+def test_or_argegarion(all_values):
+    for values1 in [{'guid_0', 'guid_1'}, {'guid_3', 'guid_6', 'guid_20'}, {'guid_2'}]:
+        for values2 in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_9'}, {'guid_2'}]:
+            test_predicate = in_reduce(
+                [in_set(values1, 'volume_guid'), in_set(values2, 'volume_guid')], any)
             included_values = set()
-            for val in PredicatesTest.all_values:
+            for val in all_values:
                 if test_predicate.do_include({'volume_guid': val}):
                     included_values.add(val)
-            self.assertEqual(included_values, PredicatesTest.all_values.difference(values))
-
-    def test_and_argegarion(self):
-        for values1 in [{'guid_0', 'guid_1'}, {'guid_3', 'guid_6', 'guid_20'}, {'guid_2'}]:
-            for values2 in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_9'}, {'guid_2'}]:
-                test_predicate = in_reduce(
-                    [in_set(values1, 'volume_guid'), in_set(values2, 'volume_guid')], all)
-                included_values = set()
-                for val in PredicatesTest.all_values:
-                    if test_predicate.do_include({'volume_guid': val}):
-                        included_values.add(val)
-                self.assertEqual(included_values, values1.intersection(values2))
-
-    def test_or_argegarion(self):
-        for values1 in [{'guid_0', 'guid_1'}, {'guid_3', 'guid_6', 'guid_20'}, {'guid_2'}]:
-            for values2 in [{'guid_2', 'guid_1'}, {'guid_5', 'guid_9'}, {'guid_2'}]:
-                test_predicate = in_reduce(
-                    [in_set(values1, 'volume_guid'), in_set(values2, 'volume_guid')], any)
-                included_values = set()
-                for val in PredicatesTest.all_values:
-                    if test_predicate.do_include({'volume_guid': val}):
-                        included_values.add(val)
-                self.assertEqual(included_values, values1.union(values2))
-
-    def test_pseudorandom_split(self):
-        split_list = [0.3, 0.4, 0.1, 0.0, 0.2]
-        values_num = len(PredicatesTest.all_values)
-        for idx in range(len(split_list)):
-            test_predicate = in_pseudorandom_split(split_list, idx, 'volume_guid')
-            included_values = set()
-            for val in PredicatesTest.all_values:
-                if test_predicate.do_include({'volume_guid': val}):
-                    included_values.add(val)
-            expected_num = values_num * split_list[idx]
-            self.assertAlmostEqual(len(included_values), expected_num, delta=expected_num * 0.1)
+            assert included_values == values1.union(values2)
 
 
-if __name__ == '__main__':
-    # Delegate to the test framework.
-    unittest.main()
+def test_pseudorandom_split(all_values):
+    split_list = [0.3, 0.4, 0.1, 0.0, 0.2]
+    values_num = len(all_values)
+    for idx in range(len(split_list)):
+        test_predicate = in_pseudorandom_split(split_list, idx, 'volume_guid')
+        included_values = set()
+        for val in all_values:
+            if test_predicate.do_include({'volume_guid': val}):
+                included_values.add(val)
+        expected_num = values_num * split_list[idx]
+        assert pytest.approx(len(included_values), expected_num * 0.1) == expected_num
+
+
+def test_predicate_on_single_column(synthetic_dataset):
+    reader = Reader(synthetic_dataset.url,
+                    schema_fields=[TestSchema.id2],
+                    predicate=in_lambda(['id2'], lambda id2: True),
+                    reader_pool=DummyPool())
+    assert len(reader) == len(synthetic_dataset.data)
+    counter = 0
+    for row in reader:
+        counter += 1
+        actual = dict(row._asdict())
+        assert actual['id2'] < 2
+    assert counter == len(synthetic_dataset.data)

